File: scripts/android-dev/app/src/main/java/com/massifmaps/test/demo/DemoOrientation.java
package com.massifmaps.test.demo;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.util.Log;

import com.massifmaps.ui.MapView;

/**
 * Points the camera where the device points: turning the phone turns the view, raising it looks up.
 *
 * This is the star-sky demo's reason for a negative tilt. The map's rotation is the opposite of the
 * heading (rotating the map right turns the view left), and the elevation above the horizon IS the
 * negative tilt the SDK now supports - at tilt -90 the view looks straight at the zenith.
 *
 * A rotation-vector sensor is fused and already smooth; the extra low pass here only takes the last
 * of the jitter out, and an update below the threshold is dropped so a still phone does not
 * re-render for ever.
 */
public final class DemoOrientation implements SensorEventListener {

    private static final String TAG = "DemoOrientation";
    /** Smoothing of the fused reading: 1 = raw, smaller = calmer and laggier. */
    private static final float SMOOTHING = 0.2f;
    /** Below this much movement, in degrees, the camera is left alone. */
    private static final float DEAD_ZONE_DEGREES = 0.3f;

    private final Context context;
    private final MapView mapView;
    private final float[] rotationMatrix = new float[9];
    private final float[] remapped = new float[9];
    private final float[] orientation = new float[3];

    private SensorManager sensorManager;
    private boolean running;
    private boolean primed;
    private float heading;
    private float elevation;

    public DemoOrientation(Context context, MapView mapView) {
        this.context = context;
        this.mapView = mapView;
    }

    public boolean isRunning() {
        return running;
    }

    public void start() {
        if (running) {
            return;
        }
        sensorManager = (SensorManager) context.getSystemService(Context.SENSOR_SERVICE);
        Sensor sensor = sensorManager != null ? sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR) : null;
        if (sensor == null) {
            Log.w(TAG, "no rotation vector sensor: orientation following is not available");
            return;
        }
        sensorManager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_GAME);
        running = true;
        primed = false;
    }

    public void stop() {
        if (!running) {
            return;
        }
        sensorManager.unregisterListener(this);
        running = false;
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        SensorManager.getRotationMatrixFromVector(rotationMatrix, event.values);
        // The phone is held up like a window on the sky, so the axis pointing OUT of the back of
        // the device is what aims the camera: remapping X/Z gives an orientation whose pitch is
        // measured from the horizon, 0 looking level and -90 looking at the zenith.
        SensorManager.remapCoordinateSystem(rotationMatrix, SensorManager.AXIS_X, SensorManager.AXIS_Z, remapped);
        SensorManager.getOrientation(remapped, orientation);

        float newHeading = (float) Math.toDegrees(orientation[0]);
        // pitch: 0 level, negative looking up. A negative map tilt is exactly the same thing.
        float newElevation = (float) Math.toDegrees(orientation[1]);

        if (!primed) {
            heading = newHeading;
            elevation = newElevation;
            primed = true;
        } else {
            // Headings wrap: smooth the SHORTEST way round, or crossing north swings the view
            // through a full turn.
            float delta = ((newHeading - heading + 540f) % 360f) - 180f;
            heading += SMOOTHING * delta;
            elevation += SMOOTHING * (newElevation - elevation);
        }

        float tilt = Math.max(-DemoConfig.LOOK_UP_LIMIT, Math.min(90f, elevation));
        float rotation = -heading;
        if (Math.abs(tilt - mapView.getTilt()) < DEAD_ZONE_DEGREES
                && Math.abs(((rotation - mapView.getMapRotation() + 540f) % 360f) - 180f) < DEAD_ZONE_DEGREES) {
            return;
        }
        mapView.setTilt(tilt, 0);
        mapView.setMapRotation(rotation, 0);
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }
}

File: scripts/ios-dev/MassifDemo/DemoOrientation.m
#import "DemoOrientation.h"
#import "DemoConfig.h"
#import <CoreMotion/CoreMotion.h>

/** Smoothing of the fused reading: 1 = raw, smaller = calmer and laggier. */
static const float SMOOTHING = 0.2f;
/** Below this much movement, in degrees, the camera is left alone. */
static const float DEAD_ZONE_DEGREES = 0.3f;

@implementation DemoOrientation {
    __weak MSFMapView *_mapView;
    CMMotionManager *_motionManager;
    BOOL _running;
    BOOL _primed;
    float _heading;
    float _elevation;
}

- (instancetype)initWithMapView:(MSFMapView *)mapView {
    if ((self = [super init])) {
        _mapView = mapView;
    }
    return self;
}

- (BOOL)running {
    return _running;
}

- (void)start {
    if (_running) {
        return;
    }
    if (!_motionManager) {
        _motionManager = [[CMMotionManager alloc] init];
    }
    if (!_motionManager.isDeviceMotionAvailable) {
        NSLog(@"MassifDemo: no device motion here (expected on the simulator)");
        return;
    }
    _running = YES;
    _primed = NO;
    _motionManager.deviceMotionUpdateInterval = 1.0 / 60.0;
    __weak DemoOrientation *weakSelf = self;
    // True north rather than magnetic, so the heading matches the map's north.
    [_motionManager startDeviceMotionUpdatesUsingReferenceFrame:CMAttitudeReferenceFrameXTrueNorthZVertical
                                                        toQueue:[NSOperationQueue mainQueue]
                                                    withHandler:^(CMDeviceMotion *motion, NSError *error) {
        [weakSelf handleMotion:motion];
    }];
}

- (void)stop {
    if (!_running) {
        return;
    }
    [_motionManager stopDeviceMotionUpdates];
    _running = NO;
}

- (void)handleMotion:(CMDeviceMotion *)motion {
    MSFMapView *mapView = _mapView;
    if (!motion || !mapView || !_running) {
        return;
    }

    // The phone is held up like a window on the sky, so the axis pointing OUT of the back of the
    // device is what aims the camera - the same thing Android's AXIS_X/AXIS_Z remap produces. The
    // attitude matrix takes a device vector into the reference frame, so the back of the device,
    // (0, 0, -1), is minus its third column.
    CMRotationMatrix m = motion.attitude.rotationMatrix;
    double x = -m.m13, y = -m.m23, z = -m.m33;
    // Reference frame: X is true north, Z is up, so Y is west - hence the negated Y for east.
    float newHeading = (float)(atan2(-y, x) * 180.0 / M_PI);
    // 0 level, negative looking up: a negative map tilt is exactly the same thing.
    float newElevation = (float)(-asin(fmax(-1.0, fmin(1.0, z))) * 180.0 / M_PI);

    if (!_primed) {
        _heading = newHeading;
        _elevation = newElevation;
        _primed = YES;
    } else {
        // Headings wrap: smooth the SHORTEST way round, or crossing north swings the view through
        // a full turn.
        float delta = fmodf(newHeading - _heading + 540.0f, 360.0f) - 180.0f;
        _heading += SMOOTHING * delta;
        _elevation += SMOOTHING * (newElevation - _elevation);
    }

    float tilt = fmaxf(-[DemoConfig floatFor:@"lookUp"], fminf(90.0f, _elevation));
    float rotation = -_heading;
    if (fabsf(tilt - [mapView getTilt]) < DEAD_ZONE_DEGREES
            && fabsf(fmodf(rotation - [mapView getRotation] + 540.0f, 360.0f) - 180.0f) < DEAD_ZONE_DEGREES) {
        return;
    }
    [mapView setTilt:tilt durationSeconds:0];
    [mapView setRotation:rotation durationSeconds:0];
}

@end
